consensus/core/src/dag_state.rs

```diff
@@ -83,27 +86,273 @@ impl DagState {
         self.cached_refs[block_ref.author].insert(block_ref);
     }
 
-    /// Gets a copy of an uncommitted block. Returns None if not found.
-    /// Uncommitted must be in memory, so only in-memory blocks are checked.
+    /// Gets an uncommitted block. Returns None if not found.
+    /// Uncommitted blocks must exist in memory, so only in-memory blocks are checked.
     pub(crate) fn get_uncommitted_block(&self, reference: &BlockRef) -> Option<VerifiedBlock> {
         self.recent_blocks.get(reference).cloned()
     }
 
-    pub(crate) fn get_blocks_at_slot(&self, _slot: Slot) -> Vec<VerifiedBlock> {
-        unimplemented!()
+    /// Gets all uncommitted blocks in a slot.
+    /// Uncommitted blocks must exist in memory, so only in-memory blocks are checked.
+    pub(crate) fn get_uncommitted_blocks_at_slot(&self, slot: Slot) -> Vec<VerifiedBlock> {
+        let mut blocks = vec![];
+        for (block_ref, block) in self.recent_blocks.range((
+            Included(BlockRef::new(slot.round, slot.authority, BlockDigest::MIN)),
+            Included(BlockRef::new(slot.round, slot.authority, BlockDigest::MAX)),
+        )) {
+            blocks.push(block.clone())
+        }
+        blocks
+    }
+
+    pub(crate) fn get_uncommitted_blocks_at_round(&self, round: Round) -> Vec<VerifiedBlock> {
+        if round < self.round_lower_bound() {
+            panic!("Round {} blocks may not be cached in memory!", round);
+        }
+
+        let mut blocks = vec![];
+        for (block_ref, block) in self.recent_blocks.range((
+            Included(BlockRef::new(round, AuthorityIndex::ZERO, BlockDigest::MIN)),
+            Excluded(BlockRef::new(
+                round + 1,
+                AuthorityIndex::ZERO,
+                BlockDigest::MIN,
+            )),
+        )) {
+            blocks.push(block.clone())
+        }
+        blocks
     }
 
-    pub(crate) fn linked_to_round(
+    pub(crate) fn ancestors_at_uncommitted_round(
         &self,
-        _later_block: &VerifiedBlock,
-        _earlier_round: Round,
+        later_block: &VerifiedBlock,
+        earlier_round: Round,
     ) -> Vec<VerifiedBlock> {
-        unimplemented!()
+        if earlier_round < self.round_lower_bound() {
+            panic!(
+                "Round {} blocks may not be cached in memory!",
+                earlier_round
+            );
+        }
+        if earlier_round >= later_block.round() {
+            panic!(
+                "Round {} is not earlier than block {}!",
+                earlier_round,
+                later_block.reference()
+            );
+        }
+
+        let mut linked: BTreeSet<BlockRef> = later_block.ancestors().iter().cloned().collect();
+        let mut round = later_block.round() - 1;
+        while round > earlier_round {
+            let mut next_linked = BTreeSet::new();
+            for r in linked.into_iter() {
```

Comment:
Since blocks can refer to other blocks of several rounds behind due to weak links, how do we expect this method to work in this case? My understanding, and I haven't worked out too deep the logic, is that we could potentially also get references to blocks with rounds < earlier_round. Or am I missing something?